### PR TITLE
Improves relative path handling for urls

### LINKED ISSUE #84: Incorrectly resolves absolute source mapping url as relative to file
The `supportRelativeURL` function always assumes mappings are relative, even when they may be an absolute URI. 

Given:
* a **file** of `c:\dir\a.min.js`
* a **url** of `file:///C:/dir/a.js`

The result is `C:\dir\file:\C:\dir\a.js` when it should be `C:\dir\a.js`.

The issue stems from the fact node's `path.resolve` function does not interpret file URIs.

### PR DESCRIPTION
**Pull request by rbuckton**
Fixes #84.

Adds two utility functions to better manage path handling: `isURI` and `toPath`.

`isURI` - Determines whether a file path is a URI using the following rules:
- A path is a URI if it starts with an alpha character followed by one or more of either an alpha character, digit, `+`, `-`, or `-`, followed by a colon. For example:
  
  ```
  http://tempuri.org/path - Web URI
  file://host/path        - File URI for UNC path
  file:///c:/path         - File URI for DOS path
  urn:custom              - Other URI
  ```
- A path is NOT a URI if it starts with a single alpha character and a colon is, instead it is treated as a DOS path. For example:
  
  ```
  c:/path                 - DOS path
  c:\path                 - DOS path
  ```
- Any other sequence of characters is not considered a URI.

`toPath` - Tries to convert a URI to a local path. If the URI is a file URI (`file://`), it is converted into a local path format that NodeJS understands using the following rules:
- A file URI with a host name is treated as a UNC path:
  
  ```
  file://server/share         -> //host/path
  ```
- A file URI without a host name, whose first path segment is a single alpha character followed by either a colon (`:`) or pipe (`|`) is treated as a rooted DOS path:
  
  ```
  file:///c:/path         -> c:\path
  file:///c|/path         -> c:\path
  ```
- A file URI without a hostname that is not treated as a DOS path is treated as a rooted POSIX path:
  
  ```
  file:///etc/path        -> /etc/path
  ```
- Querystring (`?`) and fragments (`#`) are removed.

URI handling is designed to conform to [RFC 3986](http://tools.ietf.org/html/rfc3986), with exceptions for DOS path handling.

Also fixes the tests so that they also run on Windows.

### CODE REVIEW

File: browser-test/script.map [hunk 1]
```diff
@@ -3,7 +3,7 @@
   "file": "script.js",
   "sourceRoot": "..",
   "sources": [
-    "browser-test/script.coffee"
+    "browser-test\\script.coffee"
```
Comment (LinusU) [inline browser-test/script.map]: Does this work on unix?

Comment (rbuckton) [inline browser-test/script.map]: I didn't notice that. `script.map` is a build output, but I can revert the incidental change

Comment (rbuckton) [inline browser-test/script.map]: Actually, it looks like script.map is not excluded via the .gitignore as is the case for the other tests.

**Comment by ncalexan** [comment]
Is there more work to be done here?  This looks like a more comprehensive approach to #133; can we move this forward and close that attempt?

**Comment by ncalexan** [comment]
I need a single small commit to get better printing locally: https://github.com/ncalexan/node-source-map-support/commit/bbfcde41a4510eca98e3b93de100b8505631601b

**Comment by felixfbecker** [comment]
I don't know why `isUri` is needed. Source map paths are _always_ URIs. They can be relative, but they cannot be paths (for example no backslashes). This should be enough:
```js
        let path = url.parse(fileUri).pathname;
        // convert the file URI to a path
        // strip the trailing slash from Windows paths (indicated by a drive letter with a colon)
        if (/^\/[a-zA-Z]:\//.test(path)) {
            path = path.substr(1);
        }
```

**Comment by rbuckton** [comment]
@felixfbecker `path.resolve` and `uri.resolve` work very differently. Being able to more accurately gauge which branch to take helps to make this library better able to handle a larger number of scenarios. Just because source map paths *should always* be a URI doesn't mean that tools and developers *will always* make sure it is a valid URI. This makes the library more resilient to various inputs and hopefully reduces breaking changes.

**Comment by felixfbecker** [comment]
Can't wait to get this merged...

**Comment by felixfbecker** [comment]
What's the progress on this?

**Comment by LinusU** [comment]
@rbuckton Do you feel that this is ready to merge?

@ncalexan I saw that you are using a fork, would merging this be enough for you to go back, or is there something more that we should add?

**Comment by felixfbecker** [comment]
What I noticed is that when using an absolute file uri as sourceRoot, the file path in the stack trace is also reported as a URL, while it should be converted to a path.

**Comment by felixfbecker** [comment]
Example:
```
Error: Cannot find module './routes/investment/fundings'
    at Function.Module._resolveFilename (module.js:438:15)
    at Function.Module._load (module.js:386:25)
    at Module.require (module.js:466:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (file:///c:/Users/felix/git/Projekte/registerex/src/api/app.ts:19:1)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:456:32)
    at tryModuleLoad (module.js:415:12)
    at Function.Module._load (module.js:407:3)
    at Module.require (module.js:466:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (file:///c:/Users/felix/git/Projekte/registerex/src/test/api/routes/index.ts:4:1)
    at Module._compile (module.js:541:32)
    at Object.Module._extensions..js (module.js:550:10)
    at Module.load (module.js:456:32)
    at tryModuleLoad (module.js:415:12)
    at Function.Module._load (module.js:407:3)
    at Module.require (module.js:466:17)
```
but this could be solved in a different PR, my team is currently depending on this from your fork and it works fine so far. Would really like to see this get merged.

**Comment by ncalexan** [comment]
> @rbuckton Do you feel that this is ready to merge?
> 
> @ncalexan I saw that you are using a fork, would merging this be enough for you to go back, or is there something more that we should add?

Hi folks, sorry I've been out of this loop.  The project consuming this (https://github.com/mozilla/tofino) may not really need the fork anymore; I'd need to evaluate.  My guess is we'd revert to mainline immediately after merge, and re-open if necessary.  We're not interested in maintaining forks long term.

**Comment by felixfbecker** [comment]
So can we merge this?

**Comment by rbuckton** [comment]
I have no further changes at this time, so I feel it is ready to merge.

File: test.js [hunk 2]
```diff
@@ -73,8 +74,31 @@ function createMultiLineSourceMapWithSourcesContent() {
...
+  }
+  return sourceMap;
+}
+
+
+function fileURL(file) {
+    file = path.resolve(file);
+    file = file.replace(/\\/g, '/');
+    file = file[0] !== '/' ? '/' + file : file;
+    return encodeURI('file://' + file);
```
Comment (LinusU) [inline test.js]: Can you update this (and the other new functions) to follow the 2 space indentation that the rest of the project uses?

File: test.js [hunk 3]
```diff
@@ -110,7 +147,7 @@ function compareStdout(done, sourceMap, source, expected) {
       compareLines(
         (stdout + stderr)
           .trim()
-          .split('\n')
+          .split(/\r\n|\r|\n/g)
```
Comment (LinusU) [inline test.js]: Standalone `\r` is only used on Mac OS classic which no one uses today. How about `/\r?\n/g`?

**Comment by felixfbecker** [comment]
ping @rbuckton 
can you adress the comments? Otherwise I can do it and open a new PR.

**Comment by felixfbecker** [comment]
Why close this?

**Comment by rbuckton** [comment]
It's been over two years since I opened it and I haven't had the time to update the PR or determine if it is still needed.

**Pull request closed without merging**